File: 509-fibonacci-number/509-fibonacci-number.py
class Solution:
    def __init__(self):
        self.dp = {}
    def fib(self, n: int) -> int:
        if n <= 1:
            return n
        if n not in self.dp:
            self.dp[n] = self.fib(n-1) + self.fib(n-2)
        return self.dp[n]
         

File: 509-fibonacci-number/509-fibonacci-number.java
class Solution {
    public int fib(int n) {
        int a = 0;
        int b = 1;
        int ans = 0;
        if(n == 0 || n == 1)
            return n;
        
        for(int i=1; i<n; i++) {
            ans = a + b;
            a = b;
            b = ans;
        }
        return ans;
    }
}